Library/Homebrew/cmd/fetch.rb

```diff
@@ -176,81 +218,106 @@ def run
+          previous_pending_line_count = 0
+
+          begin
+            $stdout.print Tty.hide_cursor
+            $stdout.flush
+
+            output_message = lambda do |downloadable, future|
+              status = case future.state
+              when :fulfilled
+                "#{Tty.green}✔︎#{Tty.reset}"
+              when :rejected
+                "#{Tty.red}✘#{Tty.reset}"
+              when :pending, :processing
+                "#{Tty.blue}#{spinner}#{Tty.reset}"
+              else
+                raise future.state.to_s
+              end
 
-      def fetch_resource(resource)
-        puts "Resource: #{resource.name}"
-        fetch_fetchable resource
-      rescue ChecksumMismatchError => e
-        retry if retry_fetch?(resource)
-        opoo "Resource #{resource.name} reports different sha256: #{e.expected}"
-      end
+              message = "#{downloadable.download_type.capitalize} #{downloadable.name}"
+              $stdout.puts "#{status} #{message}"
+              $stdout.flush
 
-      def fetch_formula(formula)
-        fetch_fetchable(formula)
-      rescue ChecksumMismatchError => e
-        retry if retry_fetch?(formula)
-        opoo "Formula reports different sha256: #{e.expected}"
-      end
+              if future.rejected? && (e = future.reason).is_a?(ChecksumMismatchError)
+                opoo "#{downloadable.download_type.capitalize} reports different checksum: #{e.expected}"
+                Homebrew.failed = true if downloadable.is_a?(Resource::Patch)
+                next 2
+              end
 
-      def fetch_cask(cask_download)
-        fetch_fetchable(cask_download)
-      rescue ChecksumMismatchError => e
-        retry if retry_fetch?(cask_download)
-        opoo "Cask reports different sha256: #{e.expected}"
-      end
+              1
+            end
 
-      def fetch_patch(patch)
-        fetch_fetchable(patch)
-      rescue ChecksumMismatchError => e
-        opoo "Patch reports different sha256: #{e.expected}"
-        Homebrew.failed = true
-      end
+            until remaining_downloads.empty?
+              begin
+                finished_states = [:fulfilled, :rejected]
 
-      def retry_fetch?(formula)
-        @fetch_tries ||= Hash.new { |h, k| h[k] = 1 }
-        if args.retry? && (@fetch_tries[formula] < FETCH_MAX_TRIES)
-          wait = 2 ** @fetch_tries[formula]
-          remaining = FETCH_MAX_TRIES - @fetch_tries[formula]
-          what = Utils.pluralize("tr", remaining, plural: "ies", singular: "y")
+                finished_downloads, remaining_downloads = remaining_downloads.partition do |_, future|
+                  finished_states.include?(future.state)
+                end
 
-          ohai "Retrying download in #{wait}s... (#{remaining} #{what} left)"
-          sleep wait
+                finished_downloads.each do |downloadable, future|
+                  previous_pending_line_count -= 1
+                  print "\033[K"
```

Comment:
```suggestion
                  # some comment explaining what this character is and why we print it
                  print "\033[K"
```

or may be worth a new `Formatter` function here

Reply:
```suggestion
                  # Clears to the end of the line
                  print "\033[K"
```

Reply:
Yeah, this should be in `TTY`.